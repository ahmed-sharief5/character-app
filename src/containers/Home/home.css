.home__container__characters--layout{
    background-color: #000000;
}
.home__container__characters--img--title{
    color: #ffffff;
    margin: 10px;
    position:relative;
    display:inline-block;
}
.home__container__characters--img--img {
    width: 258px;
    height: 270px;
    border-top-right-radius: 5px;
    border-top-left-radius: 5px;
}
.home__container__characters--img--title > span {
    position:absolute;
    top:72%;
    left:0;
    display:inline-block;
    background: rgba(0, 0, 0, 0.5);
    backdrop-filter: blur(1px);
    width: 258px;
    padding: 10px 5px 0px 10px;
}
.home__container__characters--img--title > span > p {
    font-size: 12px;
    color: #adadad;
}
.home__container__characters__features--section{
    background-color: #333333;
    margin: 10px;
    margin-top: -10px;
    padding: 30px 5px;
    border-bottom-left-radius: 5px;
    border-bottom-right-radius: 5px;
}
.home__container__characters__features--each-left{
    color: #a7a7a7;
    text-transform: uppercase;
    word-spacing:9999px;
}
.home__container__characters__features--each-right{
    float: right;
    color: #c17c23;
    text-transform: capitalize;
}
.home__container__characters__features--each {
    border-bottom: 1px solid #3b3b3b;
    padding: 10px;
    font-size: 12px;
}
.home__container__characters--filter {
    display: inline-block;
}
.home__container__characters__features--each:last-child {
    border-bottom: none;
}
.home__container__characters--each {
    padding: 0 !important;
    margin: 0 !important;
    /* display: inline-flex; */
}
.home__container__characters--actions {
    padding: 35px;
}
#page-numbers {
    list-style: none;
    display: flex;
    justify-content: center;
}

.pagination {
    margin: 5px;
    position: relative;
    float: left;
    padding: 6px 12px;
    margin-left: -1px;
    line-height: 1.42857143;
    color: #982f55;
    text-decoration: none;
    background-color: #fff;
    border: 1px solid #ddd;
    cursor: pointer;
}
.disabled{
    opacity: 0.65; 
    cursor: not-allowed;
    pointer-events: none;
}
.active{
    background-color: #982f55;
    color: #fff;
}
.home__container__characters__status{
    color: #ffffff;
    text-align: center;
}



   