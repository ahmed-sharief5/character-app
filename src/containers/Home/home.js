import React, { useState, useEffect } from "react";
import Filter from "../../components/Filter";
import Search from "../../components/Search";
import Sort from "../../components/Sort";
import "./home.css";
import { getCharacters } from "../../services/";
import { useSelector, useDispatch } from "react-redux";
import { fetchingCharacters, setAllCharacters } from "../../store/actions/action";
import Skeleton from "react-loading-skeleton";

const Home = () => { 
    const pageNumbersPerClick = 5;  
    const [searchedText, setSearchedText] = useState("");
    const [sortBy, setSortBy] = useState(true)
    const [selectSpeciesGender, setSelectedSpeciesGender] = useState({"species": "", "gender": ""});
    const [activeClass, setActiveClass] = useState("active");
    const [currentPagination, setcurrentPagination] = useState(0);
    const [activePage, setActivePage] = useState(1);
    const [disableClass, setDisableClass] = useState("disabled");

    const { characters, fetchingDone, error, species, gender, totalPages } = useSelector(state => state.charactersReducer);
    const dispatch = useDispatch();

    const dispatchData = (characters, species, gender, totalPages) => {
        dispatch(setAllCharacters({
            characters,
            species,
            gender,
            totalPages
        }))
    }

    async function fetchCharacters(page) {
        dispatch(fetchingCharacters({characters, species, gender, totalPages}))
        let species = [], gender = [];
        let speciesCache = {}, genderCache = {};
    
        const characters = await getCharacters(page);
        let totalPages = characters["info"]["pages"];
        if(characters["results"].length > 0){
            characters["results"].map(async character => {
                if(!speciesCache[character.species]){
                    species.push(character.species);
                    speciesCache[character.species] = 1
                }
                if(!genderCache[character.gender]){
                    gender.push(character.gender);
                    genderCache[character.gender] = 1
                }
            });
        }
        dispatchData(characters["results"],species,gender, totalPages)
    }

    const searchCharacter = (e) => {
        if(e.target.value == "")
            fetchCharacters(activePage)
        setSearchedText(e.target.value)
    }

    const _handleSearchCharacter = () => {
        let searchedCharacters = characters.filter(character => character.name.toLocaleLowerCase().includes(searchedText.toLocaleLowerCase()))
        dispatchData(searchedCharacters, species, gender, totalPages)
    }

    const _handleSortCharacter = () => {
        setSortBy(!sortBy);
        let sortedCharacters = characters.sort((a,b) => sortBy ? b.id - a.id : a.id - b.id );
        dispatchData(sortedCharacters, species, gender, totalPages)
    }

    const _handleFilterSpecies = (e, kind) => {
        e.preventDefault();
        let filteredCharacters = characters.filter(character => character[kind] === e.target.text);
        dispatchData(filteredCharacters, species, gender, totalPages);
        setSelectedSpeciesGender({...selectSpeciesGender, [kind]:e.target.text});
    }

    const _handleClearFilter = () => {
        setSelectedSpeciesGender({"species": "", "gender": ""});
        fetchCharacters(activePage);
    }

    const formatDate = (date) => {
        const formatDt = new Intl.DateTimeFormat('en-US',{year: 'numeric'}).format(new Date(date));
        return new Date().getFullYear() - new Date(formatDt).getFullYear();
    }

    // Handling the data items display
    const handleClickPage = e => {
        setActiveClass("active");
        setActivePage(Number(e.target.id));
        fetchCharacters(Number(e.target.id));
    }

    // Handling the page number display
    const handleClickPrevNext = e => {
        setActivePage(Number(e.target.id) + 1);
        setDisableClass(e.target.id <= 0 ? "disabled" : "");
        setcurrentPagination(Number(e.target.id));
    }
    
    // Logic for displaying pageNumbers
    const pageNumbers = [];
    for (let i = 1; i <= totalPages; i++) {
        pageNumbers.push(i);
    }

    const indexOfLastItemPage = currentPagination + pageNumbersPerClick;
    const indexOfFirstItemPage = indexOfLastItemPage - pageNumbersPerClick;
    const currentPageNumber = pageNumbers.slice(indexOfFirstItemPage, indexOfLastItemPage);

    const renderPageNumbers = currentPageNumber.map(number => {
    return (
        <span key={number} id={number} className={activePage === number ? `pagination ${activeClass}` : `pagination`} onClick={handleClickPage}>{number}</span>
        );
    });

    useEffect(() => {  
        fetchCharacters(activePage)
    }, [])

    return ( 
        <div className="containter">
            <div className="home__container__characters--actions container">
                <div className="row">
                    <div className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each">
                        <Sort onClick={_handleSortCharacter} byWhich="Id" sortBy={sortBy ? "Desc" : "Asc"}/>
                    </div>
                    <div className="col-lg-1 col-md-1 col-sm-1 col-xs-1 home__container__characters--each">
                        <h5>Filter By</h5>
                    </div>
                    <div className="col-lg-5 col-md-5 col-sm-11 col-xs-11 home__container__characters--each">
                        <div className="col-lg-4 col-md-4 col-sm-12 col-xs-12 home__container__characters--filter">
                            <Filter elements={species} kind={"Species"} value={selectSpeciesGender["species"]} onSelect={e => _handleFilterSpecies(e, "species")}/>
                        </div>
                        <div className="col-lg-4 col-md-4 col-sm-12 col-xs-12 home__container__characters--filter">
                            <Filter elements={gender} kind="Gender" value={selectSpeciesGender["gender"]} onSelect={e => _handleFilterSpecies(e, "gender")}/>
                        </div>
                        <div className="col-lg-4 col-md-4 col-sm-12 col-xs-12 home__container__characters--filter">
                            <button type="button" className="btn btn-danger" onClick={_handleClearFilter}>Clear</button>
                        </div>
                    </div>
                    <div className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each">
                        <Search onSearch={e => searchCharacter(e)} onSubmit={_handleSearchCharacter} byWhat="Name"/>
                    </div>
                </div>
            </div> 
            <div className="container">
                <ul id="page-numbers">
                    <span id={currentPagination - pageNumbersPerClick} className={`pagination ${disableClass}`} onClick={handleClickPrevNext}>&laquo; Previous</span>
                    {renderPageNumbers}
                    <span id={currentPagination + pageNumbersPerClick} className={`pagination ${currentPagination + pageNumbersPerClick >= totalPages ? "disabled" : ""}`} onClick={handleClickPrevNext}>&raquo; Next</span>
                </ul>
            </div>
            <div className="container">
                <div className="home__container__characters--layout container">
                    <div className="row">
                        {fetchingDone && characters && characters.length > 0 ? 
                            characters.map((character, index) => {
                                return (
                                    <div key={index} className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each">
                                        <div className="home__container__characters--img--title">
                                            <img className="home__container__characters--img--img" src={character.image} />
                                            <span>
                                                <h5>{character.name}</h5>
                                                <p>Id: {character.id} - <span>created {formatDate(character.created)} years ago</span></p>
                                            </span>
                                        </div>
                                        <div className="home__container__characters__features--section">
                                            <div className="home__container__characters__features--each">
                                                <span className="home__container__characters__features--each-left">status</span>
                                                <span className="home__container__characters__features--each-right">{character.status}</span>
                                            </div>
                                            <div className="home__container__characters__features--each">
                                                <span className="home__container__characters__features--each-left">species</span>
                                                <span className="home__container__characters__features--each-right">{character.species}</span>
                                            </div>
                                            <div className="home__container__characters__features--each">
                                                <span className="home__container__characters__features--each-left">gender</span>
                                                <span className="home__container__characters__features--each-right">{character.gender}</span>
                                            </div>
                                            <div className="home__container__characters__features--each">
                                                <span className="home__container__characters__features--each-left">origin</span>
                                                <span className="home__container__characters__features--each-right">{character.origin.name}</span>
                                            </div>
                                            <div className="home__container__characters__features--each">
                                                <span className="home__container__characters__features--each-left">last location</span>
                                                <span className="home__container__characters__features--each-right">{character.location.name}</span>
                                            </div>
                                        </div>
                                    </div>
                                );
                            })
                        : 
                        !error ? 
                            fetchingDone && characters.length == 0 ? <h2 className="home__container__characters__status">No Characters Found</h2> : 
                                <CharacterSkeleton/>
                            : <h2>{error}</h2>
                        }
                        
                    </div>
                </div> 
            </div> 
        </div>
    );
};

const CharacterSkeleton = () => {
    return (
            <div className="row">
                {Array(8)
                .fill()
                .map((item, index) => (
                    <div key={index} className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each">
                        <div className="home__container__characters--img--title">
                            {/* <img src="https://rickandmortyapi.com/api/character/avatar/1.jpeg" /> */}
                            <div className="home__container__characters--img--img">
                                <Skeleton animation="wave" height={270} width={258}/>
                            </div>
                            <span>
                                <h5><Skeleton animation="wave" height={24} width={235}/></h5>
                                <p><Skeleton animation="wave" height={18} width={235}/></p>
                            </span>
                        </div>
                        <div className="home__container__characters__features--section">
                            <div className="home__container__characters__features--each">
                                <Skeleton animation="wave" height={20} width={225}/>
                            </div>
                            <div className="home__container__characters__features--each">
                                <Skeleton animation="wave" height={20} width={225}/>
                            </div>
                            <div className="home__container__characters__features--each">
                                <Skeleton animation="wave" height={20} width={225}/>
                            </div>
                            <div className="home__container__characters__features--each">
                                <Skeleton animation="wave" height={20} width={225}/>
                            </div>
                            <div className="home__container__characters__features--each">
                                <Skeleton animation="wave" height={20} width={225}/>
                            </div>
                        </div>
                    </div>
                    // <div key={index} className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each">
                    //     <div className="home__container__characters--img--title">
                    //         <Skeleton height={200} widht={270}/>
                    //         <h5><Skeleton height={10} width={100} /></h5>
                    //         <p><Skeleton height={8} width={50} /></p>
                    //     </div>
                    //     <div className="home__container__characters__features--section">
                    //         <div className="home__container__characters__features--each">
                    //             <Skeleton height={20} width={100} />
                    //         </div>
                    //         <div className="home__container__characters__features--each">
                    //             <Skeleton height={20} width={100} />
                    //         </div>
                    //         <div className="home__container__characters__features--each">
                    //             <Skeleton height={20} width={100} />
                    //         </div>
                    //         <div className="home__container__characters__features--each">
                    //             <Skeleton height={20} width={100} />
                    //         </div>
                    //         <div className="home__container__characters__features--each">
                    //             <Skeleton height={20} width={100} />
                    //         </div>
                    //     </div>
                    // </div>
                ))}
            </div>
        
    );
}

export default Home;