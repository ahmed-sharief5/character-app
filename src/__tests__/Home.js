import React from "react";
import * as ReactRedux from 'react-redux';
import renderer from "react-test-renderer";
import sinon from "sinon";
import Home from "../containers/Home";

describe(`Home component`, () => {
    let spyDispatch, useSelectorStub, useDispatchStub;

    beforeAll(() => {
        spyDispatch = sinon.spy();
        useSelectorStub = sinon.stub(ReactRedux, "useSelector").returns({})
        useDispatchStub = sinon.stub(ReactRedux, "useDispatch").returns({spyDispatch})
    });

    afterAll(() => {
        useSelectorStub.restore();â€©
        useDispatchStub.restore();
    });

    it(`should render component correctly`, () => {
        const tree = renderer
            .create(<Home/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});