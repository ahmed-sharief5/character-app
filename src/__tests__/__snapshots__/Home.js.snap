// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should render component correctly 1`] = `
<div
  className="containter"
>
  <div
    className="home__container__characters--actions container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each"
      >
        <button
          className="btn btn-info"
          onClick={[Function]}
          type="button"
        >
          Sort Id by Desc
        </button>
      </div>
      <div
        className="col-lg-1 col-md-1 col-sm-1 col-xs-1 home__container__characters--each"
      >
        <h5>
          Filter By
        </h5>
      </div>
      <div
        className="col-lg-5 col-md-5 col-sm-11 col-xs-11 home__container__characters--each"
      >
        <div
          className="col-lg-4 col-md-4 col-sm-12 col-xs-12 home__container__characters--filter"
        >
          <div
            className="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              className="btn btn-secondary dropdown-toggle"
              data-toggle="dropdown"
              id="dropdownMenuButton"
              type="button"
            >
              Species
            </button>
            <div
              aria-labelledby="dropdownMenuButton"
              className="dropdown-menu"
            >
              <a
                className="dropdown-item"
                href="#"
              >
                No items
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-lg-4 col-md-4 col-sm-12 col-xs-12 home__container__characters--filter"
        >
          <div
            className="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              className="btn btn-secondary dropdown-toggle"
              data-toggle="dropdown"
              id="dropdownMenuButton"
              type="button"
            >
              Gender
            </button>
            <div
              aria-labelledby="dropdownMenuButton"
              className="dropdown-menu"
            >
              <a
                className="dropdown-item"
                href="#"
              >
                No items
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-lg-4 col-md-4 col-sm-12 col-xs-12 home__container__characters--filter"
        >
          <button
            className="btn btn-danger"
            onClick={[Function]}
            type="button"
          >
            Clear
          </button>
        </div>
      </div>
      <div
        className="col-lg-3 col-md-3 col-sm-6 col-xs-6 home__container__characters--each"
      >
        <div
          style={
            Object {
              "display": "inline-flex",
            }
          }
        >
          <input
            aria-label="Search"
            className="form-control mr-sm-2"
            onChange={[Function]}
            placeholder="Search By Name"
            type="search"
            value=""
          />
          <button
            className="btn btn-outline-success my-2 my-sm-0"
            onClick={[Function]}
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <ul
      id="page-numbers"
    >
      <span
        className="pagination disabled"
        id={-5}
        onClick={[Function]}
      >
        « Previous
      </span>
      <span
        className="pagination "
        id={5}
        onClick={[Function]}
      >
        » Next
      </span>
    </ul>
  </div>
  <div
    className="container"
  >
    <div
      className="home__container__characters--layout container"
    >
      <div
        className="row"
      >
        <h2
          className="home__container__characters__status"
        >
          Fetching Characters
        </h2>
      </div>
    </div>
  </div>
</div>
`;
